src/Foundation/NSDate.cs

```diff
@@ -29,23 +29,34 @@
 using System.Reflection;
 using System.Collections.Generic;
 using System.Runtime.InteropServices;
-
+using System.Threading;
 using ObjCRuntime;
 
 namespace Foundation {
 
 	public partial class NSDate {
-		const long NSDATE_TICKS = 631139040000000000;
+		const double NANOSECS_PER_MILLISEC = 1000000.0;
+
+		private static readonly NSCalendar calendar = new NSCalendar (NSCalendarType.Gregorian) { TimeZone = NSTimeZone.FromName ("UTC") };
+		private static readonly ThreadLocal<NSDateComponents> threadComponents = new ThreadLocal<NSDateComponents>(() => new NSDateComponents ());
```

Comment:
not a huge fan of adding `ThreadLocal` here

1. it can impact app size (post linking) as it bring other dependencies that might not be used (needs checking)

2. the `NSDate` -> `DateTime` conversion creates a local `NSDateComponents` instance. If it's good enough for it, why not create one (local) too for the `DateTime` -> `NSDate` conversion ?

Reply:
> it can impact app size (post linking) as it bring other dependencies that might not be used (needs checking)

Well, it was commented that allocating an NSDateComponents instance per cast was problematic. There's the option of using ThreadStatic, ThreadLocal and locking.

Both ThreadStatic and ThreadLocal achieve the same thing - but ThreadLocal avoids depleting the thread local storage by using some smart implementation details.

Also, ThreadStatic is harder to get right, as you need to null check every access and initialize if needed.

Is the overhead of ThreadLocal really that bad? It uses Func<T> and 2 small internal types, other than primitives. I don't expect a size-able increase - especially since this might go away if the explicit cast is never used.

Reply:
Not a conclusive answer re `ThreadLocal`, but @spouliot `NSDate` -> `DateTime` creates a local `NSComponents` because we have to initialize a completely new `NSDateComponents` to create components based on `NSDate`:

```
NSDateComponents calComponents = calendar.Components (NSCalendarUnit.Year ..., NSCalendarUnit.Calendar, d)
```

Reply:
yes :) which makes creating one(for the other conversion) symmetric in cost